cli/src/commands/upload.ts

```diff
@@ -0,0 +1,183 @@
+
+    const cliProgress = require('cli-progress');
+    const byteSize = require('byte-size');
+
+    const uploadProgress = new cliProgress.SingleBar(
+      {
+        format: '{bar} | {percentage}% | ETA: {eta_formatted} | {value_formatted}/{total_formatted}: {filename}',
+      },
+      cliProgress.Presets.shades_classic,
+    );
+
+    let totalSize = 0;
+    let sizeSoFar = 0;
+
+    let totalSizeUploaded = 0;
+    let uploadCounter = 0;
+
+    for (const asset of assetsToUpload) {
+      // Compute total size first
+      await asset.process();
+      totalSize += asset.fileSize;
+    }
+
+    uploadProgress.start(totalSize);
+    uploadProgress.update({ value_formatted: 0, total_formatted: byteSize(totalSize) });
+
+    for (const asset of assetsToUpload) {
+      uploadProgress.update({
+        filename: asset.path,
+      });
+
+      try {
+        if (options.import) {
+          await this.importAsset(asset);
+        } else {
+          await this.uploadAsset(asset, options.skipHash);
+        }
+      } catch (error) {
+        uploadProgress.stop();
+        throw error;
+      }
+
+      sizeSoFar += asset.fileSize;
+      if (!asset.skipped) {
+        totalSizeUploaded += asset.fileSize;
+        uploadCounter++;
+      }
+
+      uploadProgress.update(sizeSoFar, { value_formatted: byteSize(sizeSoFar) });
+    }
+
+    uploadProgress.stop();
+
+    if (options.import) {
+      if (this.dryRun) {
+        console.log(`Would have imported ${uploadCounter} assets (${byteSize(totalSizeUploaded)})`);
+      } else {
+        console.log(`Successfully imported ${uploadCounter} assets (${byteSize(totalSizeUploaded)})`);
+      }
+    } else {
+      if (uploadCounter === 0) {
+        console.log('All assets were already uploaded, nothing to do.');
+      } else {
+        if (this.dryRun) {
+          console.log(`Would have uploaded ${uploadCounter} assets (${byteSize(totalSizeUploaded)})`);
+        } else {
+          console.log(`Successfully uploaded ${uploadCounter} assets (${byteSize(totalSizeUploaded)})`);
+        }
+      }
+      if (options.delete) {
+        if (this.dryRun) {
+          console.log(`Would now have deleted assets, but skipped due to dry run`);
+        } else {
+          console.log('Deleting assets that have been uploaded...');
+          const deletionProgress = new cliProgress.SingleBar(cliProgress.Presets.shades_classic);
+          deletionProgress.start(crawledFiles.length);
+
+          for (const asset of assetsToUpload) {
+            if (!this.dryRun) {
+              await asset.delete();
+            }
+            deletionProgress.increment();
+          }
+          deletionProgress.stop();
+          console.log('Deletion complete');
+        }
+      }
+    }
+  }
+
+  private async uploadAsset(asset: CrawledAsset, skipHash = false) {
+    await asset.readData();
+
+    let skipUpload = false;
+    if (!skipHash) {
+      const checksum: string = await asset.hash();
```

Comment:
`: string` should no be needed, it should be implied by the `.hash()` return type.

Reply:
Done